amethyst_renderer_new/src/graph/mod.rs

```diff
@@ -0,0 +1,591 @@
+        C: Supports<Graphics> + Supports<Transfer>,
+    {
+        use std::iter::once;
+        use gfx_hal::command::RawCommandBuffer;
+        use gfx_hal::image::ImageLayout;
+        use gfx_hal::pool::RawCommandPool;
+        use gfx_hal::queue::RawCommandQueue;
+        use gfx_hal::queue::submission::RawSubmission;
+        use gfx_hal::window::FrameSync;
+
+        let ref signals = self.signals;
+        let ref finish = self.finish;
+        let ref acquire = self.acquire;
+        let count = self.passes.len();
+
+        // Start frame acquisition
+        let frame = SuperFrame::new(
+            &self.backbuffer,
+            swapchain.acquire_frame(FrameSync::Semaphore(acquire)),
+        );
+
+        // Allocate enough command buffers
+        if cbufs.len() < self.passes.len() {
+            let add = self.passes.len() - cbufs.len();
+            cbufs.append(&mut pool.allocate(add));
+        }
+
+        // Store `Semaphore`s that `Surface::present` needs to wait for
+        let mut presents: SmallVec<[_; 32]> = SmallVec::new();
+
+        // Record commands for all passes
+        self.passes.iter_mut().enumerate().for_each(|(id, pass)| {
+            // Pick buffer
+            let ref mut cbuf = cbufs[id];
+
+            // Begin writing
+            cbuf.begin();
+            // Setup
+            cbuf.set_viewports(&[viewport.clone()]);
+            cbuf.set_scissors(&[viewport.rect]);
+
+            // Record commands for pass
+            let acquires = pass.draw(cbuf, device, world, viewport.rect, frame.clone());
+
+            // If it renders to acquired image
+            let wait_acuired = if acquires {
+                // Presenting has to wait for it
+                presents.push(&signals[id]);
+                // And it should wait for acqusition
+                Some((acquire, PipelineStage::TOP_OF_PIPE))
+            } else {
+                None
+            };
+
+            // finish buffer recording
+            cbuf.finish();
+            unsafe {
+                // Submit buffer
+                queue.as_mut().submit_raw(
+                    RawSubmission {
+                        cmd_buffers: &[cbuf.clone()],
+                        wait_semaphores: &pass.depends
+                            .iter()
+                            .map(|&(id, stage)| (&signals[id], stage))
+                            .chain(wait_acuired)
+                            .collect::<SmallVec<[_; 32]>>(),
+
+                        // Each pass singnals to associated `Semaphore`
+                        signal_semaphores: &[&signals[id]],
+                    },
+                    // Signal the fence in last submission
+                    if id == count - 1 { Some(finish) } else { None },
+                );
+            }
+        });
+
+        // Present queue
+        swapchain.present(queue, &presents);
+
+        // Wait defice to finish the work
+        device.wait_for_fences(&[finish], WaitFor::All, !0);
```

Comment:
What pattern of using fences I should follow?
I was thinking of waiting for fence before rendering next frame instead of after.
I'm not sure if starting rendering next frame before previous one is complete is a good idea.

Reply:
![gpu](https://user-images.githubusercontent.com/7253845/33337249-b0b37776-d472-11e7-9b3e-88e877ce5a43.png)
I made a small figure which visualizes the issue, I think. The current implementation introduces a stall between GPU-CPU due to the wait for the fence. The general proposal is to buffer a few frames (basically following double/tripple buffering of the swapchain) to give the GPU time to catch up and reducing the waiting time on the fence. In the figure I used frame3 waits unitl frame0 has finished on the GPU and can then re-use data (in particular, command pools, query buffers, etc.) of frame0. The drawback here is that this makes synchronization a bit trickier as fences ensure that objects can be safely accessed again. So, each frame in the chain requires an own command pool for example.